src/fastpm.c

```diff
@@ -641,37 +648,83 @@ prepare_ncdm(FastPMSolver * fastpm, Parameters * prr, MPI_Comm comm)
     int every = CONF(prr->lua, every_ncdm);
-    
-    FastPMncdmInitData* nid = fastpm_ncdm_init_create(
-            CONF(prr->lua, boxsize),
-            m_ncdm, n_ncdm, 1 / CONF(prr->lua, time_step)[0] - 1, n_shell, n_side, lvk);
 
-  
-
-    size_t nc = CONF(prr->lua, nc) / every;
+    size_t nc_ncdm = CONF(prr->lua, nc) / every;
 
     if (CONF(prr->lua, nc) % every != 0) {
         fastpm_raise(-1, "TODO: check this in parameter file. ");
     }
-
-    size_t total_np_ncdm = (nc * nc * nc) * nid->n_split;
-
+    
+    //init the nid
+    FastPMncdmInitData* nid = fastpm_ncdm_init_create(
+            CONF(prr->lua, boxsize),
+            m_ncdm, n_ncdm, 1 / CONF(prr->lua, time_step)[0] - 1, n_shell, n_side, lvk); 
+    
+    size_t total_np_ncdm_sites = nc_ncdm * nc_ncdm * nc_ncdm;
+    size_t total_np_ncdm = total_np_ncdm_sites * nid->n_split;
+    
+    FastPMStore * cdm = fastpm_solver_get_species(fastpm, FASTPM_SPECIES_CDM);
+    
+    //create ncdm store for after the split. need to make first for memory order
     FastPMStore * ncdm = malloc(sizeof(FastPMStore));
     fastpm_store_init_evenly(ncdm,
           fastpm_species_get_name(FASTPM_SPECIES_NCDM),
           total_np_ncdm,
           cdm->attributes | COLUMN_MASS,
           fastpm->config->alloc_factor,
           comm);
+    
+    // create store for ncdm sites (i.e.before splitting) 
+    // (analogously to how cdm is created in solver_init)
+    FastPMStore * ncdm_sites = malloc(sizeof(FastPMStore));
+    fastpm_store_init_evenly(ncdm_sites,
+          fastpm_species_get_name(FASTPM_SPECIES_NCDM),
+          total_np_ncdm_sites,
+          cdm->attributes,                           //dont need mass col for sites
+          fastpm->config->alloc_factor,
+          comm);
 
+    fastpm_solver_add_species(fastpm, 
+                              FASTPM_SPECIES_NCDM, 
+                              ncdm_sites);         //will replace after splitting.
+    
+    //call fastpm_store_fill on ncdm to give correct qs
+    double shift0;
+    if(fastpm->config->USE_SHIFT) {
+        shift0 = fastpm->config->boxsize / nc_ncdm * 0.5;
+    } else {
+        shift0 = 0;
+    }
+    double shift[3] = {shift0, shift0, shift0};
+    
+    //fill the ncdm store to make a grid with nc_ncdm grid points in each dim
+    ptrdiff_t Nc_ncdm[3] = {nc_ncdm, nc_ncdm, nc_ncdm}; 
+    fastpm_store_fill(fastpm_solver_get_species(fastpm, FASTPM_SPECIES_NCDM), fastpm->basepm, shift, Nc_ncdm);
+    
+    //compute delta_k for ncdm
+    FastPMFloat * delta_k = pm_alloc(fastpm->basepm);
+    prepare_deltak(fastpm, fastpm->basepm, delta_k, prr, 1.0, 
+                   CONF(prr->lua, read_lineark_ncdm), 
+                   CONF(prr->lua, read_powerspectrum_ncdm));
+    
+    //perform lpt
+    fastpm_solver_setup_lpt(fastpm, FASTPM_SPECIES_NCDM, delta_k, CONF(prr->lua, time_step)[0]);
+    
+    //FIX could add writing of Pncdm functionality (as in prepare_ic for m).
+    pm_free(fastpm->basepm, delta_k);
```

Comment:
There is also `write_deltak`, which writes the linear field of cdm. 

In 2LPT we initialize displacement and velocity with the same deltak.

It is possible the velocity should be initialized with a different deltak,  using a velocity transfer function rather than the density transfer function. I don't recall the final verdict on this. The person to ask about this is Pat McDonald who sometimes come down hill on Tuesday.

For ncdm this doesn't matter much -- since the initial clustered velocity has very little effect at late time.

Reply:
So should I add a write_deltak_ncdm and write_powerspectrum_ncdms to prepare_ncdm so that the code outputs these?

Also, in the para file, we have 
schema.declare{name='read_lineark', type='string', ...},
whereas,
schema.declare{name='read_powerspectrum', type='file', ...}.
Should the type be a string or file?

I'll look into the velocity transfer func stuff.

Reply:
I suspect the reason `read_lineark` is a string rather than a file is because the type 'file' will assert the file exists, and not a directory.

Reply:
Let's leave the new feature to another PR.